packages/ag-charts-community/src/widget/widgetListenerInternal.ts

```diff
@@ -91,30 +98,62 @@ export class WidgetListenerInternal {
     }
 
     private startDrag<T extends Targetable>(current: T, downEvent: MouseEvent) {
+        if (WidgetListenerInternal.isDragging) return;
+        WidgetListenerInternal.isDragging = true;
+
         const window = getWindow();
         const origin: DragOrigin = { pageX: NaN, pageY: NaN, offsetX: NaN, offsetY: NaN };
         partialAssign(['pageX', 'pageY', 'offsetX', 'offsetY'], origin, downEvent);
 
+        const mousegeneral = (generalEvent: MouseEvent) => {
+            generalEvent.stopPropagation();
+            generalEvent.stopImmediatePropagation();
+        };
+
         const mousemove = (moveEvent: MouseEvent) => {
+            moveEvent.stopPropagation();
+            moveEvent.stopImmediatePropagation();
             const dragMoveEvent = makeDragEvent('drag-move', origin, moveEvent);
             this.dispatch('drag-move', current, dragMoveEvent);
         };
 
         const mouseup = (upEvent: MouseEvent) => {
             if (upEvent.button === 0) {
-                window.removeEventListener('mousemove', mousemove);
-                window.removeEventListener('mouseup', mouseup);
+                upEvent.stopPropagation();
+                upEvent.stopImmediatePropagation();
+                window.removeEventListener('mousedown', mousegeneral, { capture: true });
+                window.removeEventListener('mouseenter', mousegeneral, { capture: true });
+                window.removeEventListener('mouseleave', mousegeneral, { capture: true });
+                window.removeEventListener('mouseout', mousegeneral, { capture: true });
+                window.removeEventListener('mouseover', mousegeneral, { capture: true });
+                window.removeEventListener('mousemove', mousemove, { capture: true });
+                window.removeEventListener('mouseup', mouseup, { capture: true });
                 const dragEndEvent = makeDragEvent('drag-end', origin, upEvent);
                 this.dispatch('drag-end', current, dragEndEvent);
+                this.endDrag(current, dragEndEvent);
             }
         };
 
-        window.addEventListener('mousemove', mousemove);
-        window.addEventListener('mouseup', mouseup);
+        window.addEventListener('mousedown', mousegeneral, { capture: true });
+        window.addEventListener('mouseenter', mousegeneral, { capture: true });
+        window.addEventListener('mouseleave', mousegeneral, { capture: true });
+        window.addEventListener('mouseout', mousegeneral, { capture: true });
+        window.addEventListener('mouseover', mousegeneral, { capture: true });
+        window.addEventListener('mousemove', mousemove, { capture: true });
+        window.addEventListener('mouseup', mouseup, { capture: true });
```

Comment:
Bit of an edge case, but if the chart is destroyed in the middle of a drag, will these event handlers leak?

Reply:
No. How could that happen? `WidgetListenerInternal.destroy()` only removes `'mousedown'` event handlers on the HTMLElement, but if a drag is currently being processed then those `'mousedown'` event handles do not matter. The `'mouseup'` listener would still be registered in the `window` object, so it'll remove the handlers when it's called.

Reply:
I understand the importance of risk mitigation, but there are two important things to remember here:

1. We agreed on the design decision that the `widgets/` directory should be as isolated and independent as possible for separation of concerns (similar to the `scene/` directory). Therefore the state of the chart is of no relevance at this point. Only `this: WidgetListenerListener`, `window` and `current: T` (which is essentially an HTMLElement) are relevant.
2. The `this: WidgetListenerListener` is responsible for allocating these handlers but those not "own" them, they are conceptually owned by the `window`. Therefore removing the handlers in `WidgetListenerListener.destroy()` is more likely to cause problems than to solve anything. We would, for example, need to make sure that they are no bugs caused by double-destruction.

The only thing that the `mouseup` handler really needs from `this` is a callback. Therefore, we must go with an approach like this which is much much simpler (and safer):

```diff
diff --git a/packages/ag-charts-community/src/widget/widgetListenerInternal.ts b/packages/ag-charts-community/src/widget/widgetListenerInternal.ts
index 164a36904..588f9f780 100644
--- a/packages/ag-charts-community/src/widget/widgetListenerInternal.ts
+++ b/packages/ag-charts-community/src/widget/widgetListenerInternal.ts
@@ -128,9 +128,8 @@ export class WidgetListenerInternal {
                 window.removeEventListener('mouseover', mousegeneral, { capture: true });
                 window.removeEventListener('mousemove', mousemove, { capture: true });
                 window.removeEventListener('mouseup', mouseup, { capture: true });
-                const dragEndEvent = makeDragEvent('drag-end', origin, upEvent);
-                this.dispatch('drag-end', current, dragEndEvent);
-                this.endDrag(current, dragEndEvent);
+                const callback: undefined | ((upEvent) => void) = getCallback(upEvent);
+                callback?.(upEvent);
             }
         };
 
```

Reply:
Fixed here: https://github.com/ag-grid/ag-charts/commit/a4709c65aa6f5dac828bf95b51b73c2c7c14a5f2

Reply:
Ok, let's roll with this then.